clients/venice-producer/src/main/java/com/linkedin/venice/producer/AbstractVeniceProducer.java

```diff
@@ -0,0 +1,445 @@
+      return FLOAT_SCHEMA;
+    } else if (object instanceof byte[] || object instanceof ByteBuffer) {
+      return BYTES_SCHEMA;
+    } else if (object instanceof Boolean) {
+      return BOOL_SCHEMA;
+    } else {
+      throw new VeniceException(
+          "Venice Producer only supports Avro objects, and primitives, found object of class: "
+              + object.getClass().toString());
+    }
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(K key, V value) {
+    String error = validateProducer();
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+    return asyncPutInternal(APP_DEFAULT_LOGICAL_TS, key, value);
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(long logicalTime, K key, V value) {
+    String error = validateProducerWithLogicalTime(logicalTime);
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+    return asyncPutInternal(logicalTime, key, value);
+  }
+
+  private CompletableFuture<DurableWrite> asyncPutInternal(long logicalTime, K key, V value) {
+    producerMetrics.recordPutRequest();
+    return CompletableFuture.supplyAsync(() -> {
+      Schema valueSchema;
+      try {
+        valueSchema = getSchemaFromObject(value);
+      } catch (Exception e) {
+        producerMetrics.recordFailedRequest();
+        throw e;
+      }
+      // Might block
+      int valueSchemaId;
+      try {
+        valueSchemaId = schemaReader.getValueSchemaId(valueSchema);
+      } catch (Exception e) {
+        valueSchemaId = SchemaData.INVALID_VALUE_SCHEMA_ID;
+      }
+      if (valueSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID) {
+        producerMetrics.recordFailedRequest();
+        throw new VeniceException(
+            "Could not find a registered schema id for schema: " + valueSchema
+                + ". This might be transient if the schema has been registered recently.");
+      }
+      final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+      final Instant sendStartTime = Instant.now();
+      final PubSubProducerCallback callback = (PubSubProduceResult produceResult, Exception exception) -> {
+        Duration sendDuration = Duration.between(sendStartTime, Instant.now());
+        if (exception == null) {
+          producerMetrics.recordSuccessfulRequestWithLatency(sendDuration.toMillis());
+          completableFuture.complete(null);
+        } else {
+          producerMetrics.recordFailedRequest();
+          completableFuture.completeExceptionally(exception);
+        }
+      };
+
+      byte[] keyBytes = keySerializer.serialize(key);
+      byte[] valueBytes = getSerializer(valueSchema).serialize(value);
+
+      if (logicalTime < 0) {
+        veniceWriter.put(keyBytes, valueBytes, valueSchemaId, callback);
+      } else {
+        veniceWriter.put(keyBytes, valueBytes, valueSchemaId, logicalTime, callback);
+      }
```

Comment:
Can we simplify this and always use the API with `logicalTime` since we pass `APP_DEFAULT_LOGICAL_TS` anyway, which is the same default that `VeniceWriter` would use?